src/HappyConnOpener.h

```diff
@@ -195,6 +191,11 @@ class HappyConnOpener: public AsyncJob
     void sendSuccess(const Comm::ConnectionPointer &conn, bool reused, const char *connKind);
     void sendFailure();
 
+    /// the total number of connection opening attempts
+    int tries() const { return ale ? ale->requestAttempts : 0; }
+    /// increments the connection opening attempts
+    void incTries() { if (ale) ale->requestAttempts++; }
```

Comment:
We could simplify by removing the 'ale' check (assuming that ALE is always non-nil) but HappyConnOpener does not have this requirement. This decision depends on whether both Tunnel and FwdState callers pass a non-nil ALE here, which is not clear because they use ALE inconsistently - sometimes checking for its presence and sometimes not.

Reply:
You are right. While TunnelStateData always has ALE, FwdState is not guaranteed to have it now and did not have it in some past cases (e.g., see commit 1ce66e2).